### PR TITLE
[EKS Prow Cluster] Add karpenter terraform module for eks-prow-build cluster

### PR DESCRIPTION
**Pull request by koksay**
- Add karpenter module, which creates related roles, sqs queue and events
- Update EKSAdmin role to be able to create above resources
- Add manually added SSO role to terraform

Make sure to update the `serviceAccount.annotations` field in the `infra/aws/terraform/prow-build-cluster/resources/karpenter/flux-hr-karpenter.yaml` file (also in `infra/aws/terraform/prow-build-cluster/resources/karpenter/prod-cluster-values`):

```bash
# You may use this command to get the arn after terraform apply:
terraform state show module.karpenter.aws_iam_role.controller\[0\] | grep " arn "
```

There will be a follow-up PR to add nodepool and nodeclass configration.

### CODE REVIEW

File: infra/aws/terraform/prow-build-cluster/locals.tf [hunk 1]
```diff
@@ -81,6 +82,27 @@ locals {
     }
   ]
 
+  karpenter_roles = [
```
Comment (bryantbiggs) [inline infra/aws/terraform/prow-build-cluster/locals.tf:85]: should we use cluster access entry instead of adding to the aws-auth configmap?

File: infra/aws/terraform/prow-build-cluster/karpenter.tf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+limitations under the License.
+*/
+
+module "karpenter" {
+  source = "terraform-aws-modules/eks/aws//modules/karpenter"
+
+  cluster_name = module.eks.cluster_name
+  create_access_entry = false
+
+  enable_irsa = true
```
Comment (bryantbiggs) [inline infra/aws/terraform/prow-build-cluster/karpenter.tf:23]: why not use EKS Pod Identity here?

File: infra/aws/terraform/prow-build-cluster/locals.tf [hunk 3]
```diff
@@ -81,6 +82,27 @@ locals {
...
+      ]
+    }
+  ]
+
+  sso_roles = [
+    {
+      rolearn  = "arn:aws:iam::468814281478:role/AWSReservedSSO_AdministratorAccess_abaef4db15a2c055"
+      username = "sso-admins"
+      groups   = [
+        "eks-cluster-admin"
```
Comment (bryantbiggs) [inline infra/aws/terraform/prow-build-cluster/locals.tf:101]: not sure what permissions this group maps to, but similar to above - would cluster access entry work here instead?

File: infra/aws/terraform/prow-build-cluster/resources/karpenter/canary-cluster-values [hunk 4]
```diff
@@ -0,0 +1,23 @@
+settings:
+  clusterName: prow-canary-cluster
+  interruptionQueue: Karpenter-prow-canary-cluster
+serviceAccount:
+  annotations:
+    eks.amazonaws.com/role-arn: "arn:aws:iam::054318140392:role/KarpenterController-20240527081538529900000002"
```
Comment (bryantbiggs) [inline infra/aws/terraform/prow-build-cluster/resources/karpenter/canary-cluster-values:6]: EKS Pod Identity would remove this hardcoded role arn mapping in code

File: infra/aws/terraform/prow-build-cluster/resources/karpenter/flux-hr-karpenter.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        limits:
+          cpu: 1
+          memory: 1Gi
+        requests:
+          cpu: 1
+          memory: 1Gi
+    serviceAccount:
+      annotations:
+        # terraform state show module.karpenter.aws_iam_role.controller\[0\] | grep " arn "
+        eks.amazonaws.com/role-arn: arn:aws:iam::468814281478:role/KarpenterController-20240527081538529900000002
```
Comment (bryantbiggs) [inline infra/aws/terraform/prow-build-cluster/resources/karpenter/flux-hr-karpenter.yaml:34]: same comment - EKS Pod Identity removes this hardcoding

File: infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf [hunk 6]
```diff
@@ -90,6 +101,12 @@ data "aws_iam_policy_document" "eks_apply" {
       "logs:PutRetentionPolicy",
       "logs:TagLogGroup",
       "s3:PutObject",
+      "sqs:createqueue",
```
Comment (bryantbiggs) [inline infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf:104]: Are these permissions required by prow, or by Karpenter? Likewise fir the Eventbridge API permissions

Comment (koksay) [reply infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf:104]: These are required by the karpenter module, it creates a SQS queue for the event messaging.

File: infra/aws/terraform/prow-build-cluster/locals.tf (hunk 1 above)
Comment (koksay) [reply infra/aws/terraform/prow-build-cluster/locals.tf:85]: That's a good question. I tried this code on the Canary cluster first and could not run Karpenter with the access entry. I may have missed something, so I turned it off and continued with the aws-auth file.

File: infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf [hunk 7]
```diff
@@ -90,6 +101,12 @@ data "aws_iam_policy_document" "eks_apply" {
...
+      "sqs:deletequeue",
+      "sqs:getqueueattributes",
+      "sqs:listqueuetags",
```
[the remark this answers is not in the document]
Comment (xmudrii) [reply infra/aws/terraform/modules/eks-prow-iam/policy_apply.tf:107]: Why are those permissions lowercase while all other permissions are CamelCase? Is that on the AWS side or AWS doesn't care at all?

File: infra/aws/terraform/prow-build-cluster/resources/karpenter/canary-cluster-values [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    requests:
+      cpu: 1
+      memory: 1Gi
+    limits:
+      cpu: 1
+      memory: 1Gi
+
+# default affinity rule works well in our case
+# https://github.com/aws/karpenter-provider-aws/blob/main/charts/karpenter/values.yaml#L70
+tolerations:
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/resources/karpenter/canary-cluster-values:19]: Do we want to force scheduling Karpenter on the stable nodes?

File: infra/aws/terraform/prow-build-cluster/resources/karpenter/prod-cluster-values [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    requests:
+      cpu: 1
+      memory: 1Gi
+    limits:
+      cpu: 1
+      memory: 1Gi
+
+# default affinity rule works well in our case
+# https://github.com/aws/karpenter-provider-aws/blob/main/charts/karpenter/values.yaml#L70
+tolerations:
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/resources/karpenter/prod-cluster-values:19]: Same here, we should check if we want to force Karpenter on the stable nodes.

File: infra/aws/terraform/prow-build-cluster/vpc.tf [hunk 10]
```diff
@@ -65,6 +64,7 @@ module "vpc" {
   public_subnet_tags = {
     "kubernetes.io/role/elb"                    = 1
     "kubernetes.io/cluster/${var.cluster_name}" = "owned"
+    "karpenter.sh/discovery"                    = var.cluster_name
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/vpc.tf:67]: Can we add a comment why is this required?

**Review by xmudrii — CHANGES_REQUESTED** [review]

**Comment by xmudrii** [comment]
/hold
Until we don't decide on https://github.com/kubernetes/k8s.io/pull/7062

@koksay I worked on this PR locally and fixed most of the comments, I'll see how can I push those changes or eventually create a new PR if that's okay
